opentasks-provider/src/main/java/org/dmfs/provider/tasks/utils/InstanceDateTimeData.java

```diff
@@ -0,0 +1,111 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.dmfs.provider.tasks.utils;
+
+import android.content.ContentValues;
+
+import org.dmfs.iterables.elementary.Seq;
+import org.dmfs.jems.function.BiFunction;
+import org.dmfs.jems.single.Single;
+import org.dmfs.optional.Optional;
+import org.dmfs.optional.adapters.FirstPresent;
+import org.dmfs.optional.composite.Zipped;
+import org.dmfs.rfc5545.DateTime;
+import org.dmfs.rfc5545.Duration;
+import org.dmfs.tasks.contract.TaskContract;
+
+import java.util.TimeZone;
+
+
+/**
+ * A {@link Single} holding the date and time {@link ContentValues} of an instance.
+ *
+ * @author Marten Gajda
+ */
+public final class InstanceDateTimeData implements Single<ContentValues>
+{
+    private final Optional<DateTime> mOptionalStart;
+    private final Optional<DateTime> mOptionalDue;
+    private final Optional<Duration> mOptionalDuration;
+    private final Optional<DateTime> mOriginalStart;
+
+
+    public InstanceDateTimeData(Optional<DateTime> optionalStart, Optional<DateTime> optionalDue, Optional<Duration> optionalDuration, Optional<DateTime> originalStart)
+    {
+        mOptionalStart = optionalStart;
+        mOptionalDue = optionalDue;
+        mOptionalDuration = optionalDuration;
+        mOriginalStart = originalStart;
+    }
+
+
+    @Override
+    public ContentValues value()
+    {
+        ContentValues instanceValues = new ContentValues();
+        TimeZone localTz = TimeZone.getDefault();
+
+        Optional<DateTime> effectiveDue = new FirstPresent<>(
+                new Seq<>(mOptionalDue, new Zipped<>(mOptionalStart, mOptionalDuration, new AddDurationBiFunction())));
+        Optional<Long> effectiveDuration = new Zipped<>(mOptionalStart, effectiveDue, new DurationBiFunction());
+
+        putDateValue(instanceValues, mOptionalStart, TaskContract.Instances.INSTANCE_START, TaskContract.Instances.INSTANCE_START_SORTING, localTz);
+        putDateValue(instanceValues, effectiveDue, TaskContract.Instances.INSTANCE_DUE, TaskContract.Instances.INSTANCE_DUE_SORTING, localTz);
+        instanceValues.put(TaskContract.Instances.INSTANCE_DURATION, effectiveDuration.value(null));
+        instanceValues.put(TaskContract.Instances.INSTANCE_ORIGINAL_TIME,
+                new FirstPresent<>(new Seq<>(mOriginalStart, mOptionalStart, effectiveDue)).value(new DateTime(0)).getTimestamp());
+
+        return instanceValues;
+    }
+
+
+    private void putDateValue(ContentValues values, Optional<DateTime> value, String timestampColumn, String sortingColumn, TimeZone localTimeZone)
+    {
+        if (value.isPresent())
+        {
+            // add timestamp and sorting
+            DateTime dateTime = value.value();
+            values.put(timestampColumn, dateTime.getTimestamp());
+            values.put(sortingColumn, dateTime.isAllDay() ? dateTime.getInstance() : dateTime.shiftTimeZone(localTimeZone).getInstance());
```

Comment:
I see here that the sorting column has different value than the timestamp one..   could you help explain a bit what is this column for? Is it related to performance optimization?

Reply:
As explained above, this is to guarantee correct sorting of floating events. Since they are stored with a UTC timestamp, they would show up at wrong dates and times when sorted by timestamp along with non-floating tasks.
I.e. consider a floating task on June 1st at 8:00 "local time" and another one on June 1st at 9:00 CET. The floating task is stored with a UTC timestamp of "June 1st, 8:00" the CET task is stored with a UTC timestamp of "June 1st 7:00". So if you're located in CET and you sort these by timestamp the 8:00 task will be shown after the 9:00 task.

The sorting columns hold a representation in "local time". So the floating task is stored as "June 1st 8:00" and the absolute task is stored as "June 1st 9:00". Sorting them will show them in the correct order.

Note: when you change the time zone of your device, we need to re-calculate the sorting columns of all tasks and instances. That's why we have this `UPDATE_REQUESTED` hack in the `Instantiating` processor. It ensures all instances are updated even if no datetime value had been modified.

Reply:
I see, thanks for explaining.